src/operations/blas3/gemm_local_joint_matrix.hpp

```diff
@@ -0,0 +1,866 @@
+   * @param scratch local memory
+   */
+  template <typename local_memory_t>
+  SYCL_BLAS_INLINE void eval(local_memory_t scratch_acc,
+                             const cl::sycl::nd_item<1> &id) noexcept {
+    index_t m = a_.get_size_row();
+    index_t n = b_.get_size_col();
+    index_t k = a_.get_size_col();
+
+    const index_t lda = a_.getSizeL();
+    const index_t ldb = b_.getSizeL();
+    const index_t ldc = c_.getSizeL();
+    // The batch index that each workgroup should start working with
+    const index_t x_groups = (get_wg_x_cluster() - 1) / jm_row_frags + 1;
+    const index_t y_groups = (get_wg_y_cluster() - 1) / jm_col_frags + 1;
+    const index_t wg_batch_id = id.get_group(0) / (x_groups * y_groups);
+    // This will disable all workgroups that dont have any batch to work on
+    if (wg_batch_id >= batch_size_) {
+      return;
+    }
+    const index_t batch_stride = id.get_group_range(0) / (x_groups * y_groups);
+
+    auto scratch = scratch_acc.localAcc.get_pointer();
+
+    // The number of work-group required to executed each batch efficiently
+    const index_t wg_id_x = id.get_group(0) % x_groups;
+    const index_t wg_id_y = (id.get_group(0) / x_groups) % y_groups;
+
+    const index_t a_size = trans_a ? m * lda : k * lda;
+    const index_t b_size = trans_b ? ldb * k : n * ldb;
+    const index_t c_size = ldc * n;
+
+    auto ptr_A = a_.get_data().get_pointer() + a_.get_access_displacement() +
+                 (wg_batch_id * a_size);
+    auto ptr_B = b_.get_data().get_pointer() + b_.get_access_displacement() +
+                 (wg_batch_id * b_size);
+    auto ptr_C = c_.get_data().get_pointer() + c_.get_access_displacement() +
+                 (wg_batch_id * c_size);
+
+    auto sg = id.get_sub_group();
+    const index_t sg_id = sg.get_group_linear_id();
+    const index_t item_id = id.get_local_linear_id();
+
+    const index_t wg_row = wg_id_x * block_rows;
+    const index_t wg_col = wg_id_y * block_cols;
+    const bool out_of_range = (wg_row >= m || wg_col >= n);
+    const bool internal = m - wg_row >= block_rows && n - wg_col >= block_cols;
+
+    ptr_C +=
+        (wg_row + (sg_id % jm_row_frags) * tile_type::joint_matrix_M) +
+        (wg_col + (sg_id / jm_row_frags) * tile_type::joint_matrix_N) * ldc;
+
+    const index_t mc = m - wg_row;
+    const index_t nc = n - wg_col;
+
+    const index_t it_mod_brows = item_id % block_rows;
+    const index_t it_div_brows = item_id / block_rows;
+
+    const index_t it_mod_cl = item_id % cl_elems;
+    const index_t it_div_cl = item_id / cl_elems;
+
+    ptr_B += (trans_b ? (item_id / block_cols) * ldb +
+                            (wg_col + item_id % block_cols)
+                      : it_mod_cl + (wg_col + it_div_cl) * ldb);
+
+    n = n - wg_col - (trans_b ? item_id % block_cols : it_div_cl);
+    ptr_A += (trans_a ? (wg_row + it_div_cl) * lda + (it_mod_cl)
+                      : (wg_row + it_mod_brows) + (item_id / block_rows) * lda);
```

Comment:
```suggestion
    const index_t it_mod_brows = item_id % block_rows;
    const index_t it_div_brows = item_id / block_rows;

    const index_t it_mod_cl = item_id % cl_elems;
    const index_t it_div_cl = item_id / cl_elems;

    ptr_B += (trans_b ? (item_id / block_cols) * ldb +
                            (wg_col + item_id % block_cols)
                      : it_mod_cl + (wg_col + it_div_cl) * ldb);

    n = n - wg_col - (trans_b ? item_id % block_cols : it_div_cl);
    ptr_A += (trans_a ? (wg_row + it_div_cl) * lda + (it_mod_cl)
                      : (wg_row + it_mod_brows) + it_div_brows * lda);
```
Why does block rows get pre-computed but not block_cols?

Reply:
Fixed this in the latest commit.